File: src-tauri/src/query/validate.rs
use std::collections::HashSet;

use crate::introspect::{ColumnCategory, Schema, classify_column};
use crate::query::ir::{FilterNode, FilterOp, LiteralKind, QuerySpec};

#[path = "validate_helpers.rs"]
mod validate_helpers;

pub const LARGE_LIMIT_WARNING_THRESHOLD: u32 = 1000;
pub const MAX_LIMIT: u32 = 10_000;
pub const DEFAULT_LIMIT: u32 = 100;
pub const MAX_FILTER_DEPTH: usize = 5;
pub const MAX_IN_LIST_SIZE: usize = 1000;
pub const MAX_TEXT_LITERAL_LEN: usize = 10_000;

const BLOCKED_SCHEMAS: &[&str] = &[
    "pg_catalog",
    "information_schema",
    "pg_toast",
    "mysql",
    "performance_schema",
    "sys",
    "guest",
    "INFORMATION_SCHEMA",
    "SYS",
    "SYSTEM",
    "OUTLN",
    "DBSNMP",
    "XDB",
    "CTXSYS",
    "MDSYS",
    "OLAPSYS",
    "WMSYS",
    "ORDSYS",
    "EXFSYS",
    "ANONYMOUS",
    "APEX_PUBLIC_USER",
    "FLOWS_FILES",
    "APEX_030200",
    "APEX_040000",
    "APEX_040200",
    "AUDSYS",
    "GSMADMIN_INTERNAL",
    "SYSMAN",
    "DBSFWUSER",
    "APPQOSSYS",
    "ORACLE_OCM",
    "XS$NULL",
    "DVSYS",
    "LBACSYS",
];

#[derive(Debug)]
pub struct ValidationOutcome {
    pub warnings: Vec<String>,
    pub limit: u32,
}

#[derive(Debug, Clone)]
pub struct ValidatedColumn {
    pub table_alias: String,
    pub column: String,
    pub result_alias: String,
}

#[derive(Debug, Clone)]
pub struct ValidatedQuery {
    spec: QuerySpec,
    columns: Vec<ValidatedColumn>,
}

impl ValidatedQuery {
    pub(crate) fn spec(&self) -> &QuerySpec {
        &self.spec
    }

    pub(crate) fn columns(&self) -> &[ValidatedColumn] {
        &self.columns
    }
}

pub fn validate_query(
    spec: &mut QuerySpec,
    schema: &Schema,
    custom_blocked: &[String],
) -> Result<(ValidatedQuery, ValidationOutcome), String> {
    let outcome = validate(spec, schema, custom_blocked)?;
    let selections = if spec.columns.is_empty() {
        spec.tables
            .iter()
            .flat_map(|table_ref| {
                find_table(schema, &table_ref.schema, &table_ref.name)
                    .into_iter()
                    .flat_map(move |table| {
                        table
                            .columns
                            .iter()
                            .map(move |column| crate::query::ir::ColumnSel {
                                table_alias: table_ref.alias.clone(),
                                column: column.name.clone(),
                            })
                    })
            })
            .collect::<Vec<_>>()
    } else {
        spec.columns.clone()
    };
    let mut aliases = HashSet::new();
    let columns = selections
        .into_iter()
        .map(|selection| {
            let base = format!("{}__{}", selection.table_alias, selection.column);
            let mut result_alias = base.clone();
            let mut suffix = 2usize;
            while !aliases.insert(result_alias.clone()) {
                result_alias = format!("{base}__{suffix}");
                suffix += 1;
            }
            ValidatedColumn {
                table_alias: selection.table_alias,
                column: selection.column,
                result_alias,
            }
        })
        .collect();
    Ok((
        ValidatedQuery {
            spec: spec.clone(),
            columns,
        },
        outcome,
    ))
}

pub fn literal_kind_for_column(data_type: &str) -> LiteralKind {
    match classify_column(data_type) {
        ColumnCategory::Integer => LiteralKind::Int,
        ColumnCategory::Decimal => LiteralKind::Decimal,
        ColumnCategory::Bool => LiteralKind::Bool,
        ColumnCategory::Date => LiteralKind::Date,
        ColumnCategory::DateTime => LiteralKind::DateTime,
        ColumnCategory::Text
        | ColumnCategory::Binary
        | ColumnCategory::Json
        | ColumnCategory::Other => LiteralKind::Text,
    }
}

pub fn ops_for_column(data_type: &str) -> &'static [FilterOp] {
    match classify_column(data_type) {
        ColumnCategory::Text => &[
            FilterOp::Eq,
            FilterOp::Ne,
            FilterOp::Like,
            FilterOp::NotLike,
            FilterOp::Ilike,
            FilterOp::In,
            FilterOp::NotIn,
            FilterOp::IsNull,
            FilterOp::IsNotNull,
            FilterOp::IsEmpty,
            FilterOp::IsNotEmpty,
        ],
        ColumnCategory::Integer | ColumnCategory::Decimal => &[
            FilterOp::Eq,
            FilterOp::Ne,
            FilterOp::Gt,
            FilterOp::Gte,
            FilterOp::Lt,
            FilterOp::Lte,
            FilterOp::In,
            FilterOp::NotIn,
            FilterOp::Between,
            FilterOp::IsNull,
            FilterOp::IsNotNull,
        ],
        ColumnCategory::Bool => &[
            FilterOp::Eq,
            FilterOp::Ne,
            FilterOp::IsNull,
            FilterOp::IsNotNull,
        ],
        ColumnCategory::Date | ColumnCategory::DateTime => &[
            FilterOp::Eq,
            FilterOp::Ne,
            FilterOp::Gt,
            FilterOp::Gte,
            FilterOp::Lt,
            FilterOp::Lte,
            FilterOp::Between,
            FilterOp::IsNull,
            FilterOp::IsNotNull,
        ],
        ColumnCategory::Binary | ColumnCategory::Json | ColumnCategory::Other => &[
            FilterOp::Eq,
            FilterOp::Ne,
            FilterOp::IsNull,
            FilterOp::IsNotNull,
        ],
    }
}

pub fn validate(
    spec: &mut QuerySpec,
    schema: &Schema,
    custom_blocked: &[String],
) -> Result<ValidationOutcome, String> {
    let mut warnings = Vec::new();

    if spec.schema_version != crate::query::ir::CURRENT_SCHEMA_VERSION {
        return Err(format!(
            "Query schema version {} is unsupported; expected {}",
            spec.schema_version,
            crate::query::ir::CURRENT_SCHEMA_VERSION
        ));
    }
    let mut node_ids = HashSet::new();
    validate_helpers::collect_node_ids(&spec.filters, &mut node_ids)?;
    for override_id in spec.connector_overrides.keys() {
        if !node_ids.contains(override_id.as_str()) {
            return Err(format!(
                "Connector override references unknown filter node id '{override_id}'"
            ));
        }
    }

    if spec.tables.is_empty() {
        return Err("At least one table is required".to_string());
    }

    let mut table_aliases: HashSet<&str> = HashSet::new();
    for table in &spec.tables {
        if validate_helpers::is_blocked(&table.schema, custom_blocked) {
            return Err(format!(
                "Schema '{}' is blocked (system/catalog schema)",
                table.schema
            ));
        }

        let _table_info = find_table(schema, &table.schema, &table.name).ok_or_else(|| {
            format!(
                "Table '{}.{}' not found in schema",
                table.schema, table.name
            )
        })?;

        if !table_aliases.insert(table.alias.as_str()) {
            return Err(format!("Duplicate table alias '{}'", table.alias));
        }

        for col in &spec.columns {
            if col.table_alias == table.alias
                && !find_table(schema, &table.schema, &table.name)
                    .map(|t| t.columns.iter().any(|c| c.name == col.column))
                    .unwrap_or(false)
            {
                return Err(format!(
                    "Column '{}.{}' does not exist",
                    table.alias, col.column
                ));
            }
        }
    }

    for col in &spec.columns {
        if !table_aliases.contains(col.table_alias.as_str()) {
            return Err(format!(
                "Column selection references unknown table alias '{}'",
                col.table_alias
            ));
        }
    }

    for join in &spec.joins {
        if !table_aliases.contains(join.left_alias.as_str()) {
            return Err(format!(
                "Join references unknown table alias '{}'",
                join.left_alias
            ));
        }
        if !table_aliases.contains(join.right_alias.as_str()) {
            return Err(format!(
                "Join references unknown table alias '{}'",
                join.right_alias
            ));
        }

        let left_table = find_table_by_alias(schema, spec, &join.left_alias)
            .ok_or_else(|| format!("Cannot resolve table for alias '{}'", join.left_alias))?;
        let right_table = find_table_by_alias(schema, spec, &join.right_alias)
            .ok_or_else(|| format!("Cannot resolve table for alias '{}'", join.right_alias))?;

        let left_col = left_table
            .columns
            .iter()
            .find(|c| c.name == join.left_column)
            .ok_or_else(|| {
                format!(
                    "Join column '{}.{}' does not exist",
                    join.left_alias, join.left_column
                )
            })?;
        let right_col = right_table
            .columns
            .iter()
            .find(|c| c.name == join.right_column)
            .ok_or_else(|| {
                format!(
                    "Join column '{}.{}' does not exist",
                    join.right_alias, join.right_column
                )
            })?;

        if !left_col.join_eligible {
            return Err(format!(
                "Join column '{}.{}' is not the leading key of an equality-capable index",
                join.left_alias, join.left_column
            ));
        }
        if !right_col.join_eligible {
            return Err(format!(
                "Join column '{}.{}' is not the leading key of an equality-capable index",
                join.right_alias, join.right_column
            ));
        }
        if left_col.category != right_col.category {
            return Err(format!(
                "Join columns '{}.{}' and '{}.{}' have incompatible types",
                join.left_alias, join.left_column, join.right_alias, join.right_column
            ));
        }
    }

    validate_helpers::validate_node(
        &FilterNode::Group(spec.filters.clone()),
        schema,
        spec,
        &table_aliases,
        0,
        &mut warnings,
    )?;

    if spec.tables.len() > 1 && !validate_helpers::check_join_connectivity(spec) {
        return Err(
            "Not all tables are connected by joins — add joins linking every table".to_string(),
        );
    }

    if spec.limit == 0 {
        spec.limit = DEFAULT_LIMIT;
        warnings.push(format!("Limit was 0; defaulted to {}", DEFAULT_LIMIT));
    }
    if spec.limit > MAX_LIMIT {
        warnings.push(format!(
            "Limit {} exceeds maximum {}; capped to {}",
            spec.limit, MAX_LIMIT, MAX_LIMIT
        ));
        spec.limit = MAX_LIMIT;
    }
    if spec.limit > LARGE_LIMIT_WARNING_THRESHOLD {
        warnings.push(
            "Large result limits are useful for reporting, but filters, selected columns, and indexed predicates keep queries faster and easier to reuse."
                .to_string(),
        );
    }

    if spec.columns.is_empty() {
        warnings.push("No columns selected — query will select all columns".to_string());
    }

    Ok(ValidationOutcome {
        warnings,
        limit: spec.limit,
    })
}

fn find_table<'a>(
    schema: &'a Schema,
    table_schema: &str,
    table_name: &str,
) -> Option<&'a crate::introspect::TableInfo> {
    schema
        .tables
        .iter()
        .find(|t| t.schema == table_schema && t.name == table_name)
}

fn find_table_by_alias<'a>(
    schema: &'a Schema,
    spec: &QuerySpec,
    alias: &str,
) -> Option<&'a crate::introspect::TableInfo> {
    let table_ref = spec.tables.iter().find(|t| t.alias == alias)?;
    find_table(schema, &table_ref.schema, &table_ref.name)
}

File: shared/src/main/kotlin/com/safedb/query/Validate.kt
package com.safedb.query

import com.safedb.model.CURRENT_SCHEMA_VERSION
import com.safedb.model.ColumnCategory
import com.safedb.model.ColumnSel
import com.safedb.model.Dialect
import com.safedb.model.FilterNode
import com.safedb.model.FilterOp
import com.safedb.model.JoinSpec
import com.safedb.model.LiteralKind
import com.safedb.model.Outcome
import com.safedb.model.QuerySpec
import com.safedb.model.Schema
import com.safedb.model.SortSpec
import com.safedb.model.classifyColumn

const val LARGE_LIMIT_WARNING_THRESHOLD = 1000
const val MAX_LIMIT = 5_000
const val DEFAULT_LIMIT = 100
const val MAX_FILTER_DEPTH = 5
const val MAX_IN_LIST_SIZE = 1000
const val MAX_TEXT_LITERAL_LEN = 10_000

internal val BLOCKED_SCHEMAS =
    listOf(
        "pg_catalog",
        "information_schema",
        "pg_toast",
        "mysql",
        "performance_schema",
        "sys",
        "guest",
        "INFORMATION_SCHEMA",
        "SYS",
        "SYSTEM",
        "OUTLN",
        "DBSNMP",
        "XDB",
        "CTXSYS",
        "MDSYS",
        "OLAPSYS",
        "WMSYS",
        "ORDSYS",
        "EXFSYS",
        "ANONYMOUS",
        "APEX_PUBLIC_USER",
        "FLOWS_FILES",
        "APEX_030200",
        "APEX_040000",
        "APEX_040200",
        "AUDSYS",
        "GSMADMIN_INTERNAL",
        "SYSMAN",
        "DBSFWUSER",
        "APPQOSSYS",
        "ORACLE_OCM",
        $$"XS$NULL",
        "DVSYS",
        "LBACSYS",
    )

data class ValidationOutcome(val warnings: List<String>, val limit: Int)

data class ValidatedColumn(val tableAlias: String, val column: String, val resultAlias: String)

class ValidatedQuery
internal constructor(private val spec: QuerySpec, private val columns: List<ValidatedColumn>) {
    fun spec(): QuerySpec = spec

    fun columns(): List<ValidatedColumn> = columns
}

fun validateQuery(
    spec: QuerySpec,
    schema: Schema,
    customBlocked: List<String>,
    dialect: Dialect? = null,
): Outcome<Pair<ValidatedQuery, ValidationOutcome>> {
    val (normalizedSpec, outcome) =
        when (val result = validate(spec, schema, customBlocked, dialect)) {
            is Outcome.Ok -> result.value
            is Outcome.Err -> return Outcome.err(result.message)
        }

    val selections =
        normalizedSpec.columns.ifEmpty {
            normalizedSpec.tables.flatMap { tableRef ->
                findTable(schema, tableRef.schema, tableRef.name)?.columns.orEmpty().map { column ->
                    ColumnSel(tableAlias = tableRef.alias, column = column.name)
                }
            }
        }

    val aliases = mutableSetOf<String>()
    val validatedColumns = selections.map { selection ->
        val base = "${selection.tableAlias}__${selection.column}"
        var resultAlias = base
        var suffix = 2
        while (!aliases.add(resultAlias)) {
            resultAlias = "${base}__$suffix"
            suffix++
        }
        ValidatedColumn(
            tableAlias = selection.tableAlias,
            column = selection.column,
            resultAlias = resultAlias,
        )
    }

    return Outcome.ok(ValidatedQuery(normalizedSpec, validatedColumns) to outcome)
}

fun literalKindForColumn(dataType: String): LiteralKind =
    when (classifyColumn(dataType)) {
        ColumnCategory.Integer -> LiteralKind.Int
        ColumnCategory.Decimal -> LiteralKind.Decimal
        ColumnCategory.Bool -> LiteralKind.Bool
        ColumnCategory.Date -> LiteralKind.Date
        ColumnCategory.DateTime -> LiteralKind.DateTime
        ColumnCategory.Text,
        ColumnCategory.Binary,
        ColumnCategory.Json,
        ColumnCategory.Other -> LiteralKind.Text
    }

fun opsForColumn(dataType: String): List<FilterOp> =
    when (classifyColumn(dataType)) {
        ColumnCategory.Text ->
            listOf(
                FilterOp.Eq,
                FilterOp.Ne,
                FilterOp.Contains,
                FilterOp.ContainsIgnoreCase,
                FilterOp.NotContains,
                FilterOp.StartsWith,
                FilterOp.EndsWith,
                FilterOp.In,
                FilterOp.NotIn,
                FilterOp.IsNull,
                FilterOp.IsNotNull,
                FilterOp.IsEmpty,
                FilterOp.IsNotEmpty,
            )
        ColumnCategory.Integer,
        ColumnCategory.Decimal ->
            listOf(
                FilterOp.Eq,
                FilterOp.Ne,
                FilterOp.Gt,
                FilterOp.Gte,
                FilterOp.Lt,
                FilterOp.Lte,
                FilterOp.In,
                FilterOp.NotIn,
                FilterOp.Between,
                FilterOp.IsNull,
                FilterOp.IsNotNull,
            )
        ColumnCategory.Bool -> listOf(FilterOp.Eq, FilterOp.Ne, FilterOp.IsNull, FilterOp.IsNotNull)
        ColumnCategory.Date,
        ColumnCategory.DateTime ->
            listOf(
                FilterOp.Eq,
                FilterOp.Ne,
                FilterOp.Gt,
                FilterOp.Gte,
                FilterOp.Lt,
                FilterOp.Lte,
                FilterOp.Between,
                FilterOp.IsNull,
                FilterOp.IsNotNull,
            )
        ColumnCategory.Binary,
        ColumnCategory.Json,
        ColumnCategory.Other ->
            listOf(FilterOp.Eq, FilterOp.Ne, FilterOp.IsNull, FilterOp.IsNotNull)
    }

// Includes historical raw LIKE operators so saved queries continue to validate and run.
internal fun validOpsForColumn(dataType: String): List<FilterOp> =
    opsForColumn(dataType) +
        when (classifyColumn(dataType)) {
            ColumnCategory.Text -> listOf(FilterOp.Like, FilterOp.NotLike, FilterOp.Ilike)
            else -> emptyList()
        }

fun distinctSortProjectionConflicts(spec: QuerySpec): List<SortSpec> {
    if (!spec.distinct || spec.columns.isEmpty()) return emptyList()
    val selectedColumns = spec.columns.mapTo(mutableSetOf()) { it.tableAlias to it.column }
    return spec.sorts.filter { (it.tableAlias to it.column) !in selectedColumns }
}

fun validate(
    spec: QuerySpec,
    schema: Schema,
    customBlocked: List<String>,
    dialect: Dialect? = null,
): Outcome<Pair<QuerySpec, ValidationOutcome>> {
    val warnings = mutableListOf<String>()

    if (spec.schemaVersion != CURRENT_SCHEMA_VERSION) {
        return Outcome.err(
            "Query schema version ${spec.schemaVersion} is unsupported; expected $CURRENT_SCHEMA_VERSION"
        )
    }

    val nodeIds = mutableSetOf<String>()
    when (val result = collectNodeIds(spec.filters, nodeIds)) {
        is Outcome.Ok -> Unit
        is Outcome.Err -> return Outcome.err(result.message)
    }

    for (overrideId in spec.connectorOverrides.keys) {
        if (!nodeIds.contains(overrideId)) {
            return Outcome.err("Connector override references unknown filter node id '$overrideId'")
        }
    }

    if (spec.tables.isEmpty()) {
        return Outcome.err("At least one table is required")
    }

    val tableAliases = mutableSetOf<String>()
    for ((tableSchema, tableName, alias) in spec.tables) {
        if (isBlocked(tableSchema, customBlocked)) {
            return Outcome.err("Schema '$tableSchema' is blocked (system/catalog schema)")
        }

        if (findTable(schema, tableSchema, tableName) == null) {
            return Outcome.err("Table '$tableSchema.$tableName' not found in schema")
        }

        if (!tableAliases.add(alias)) {
            return Outcome.err("Duplicate table alias '$alias'")
        }

        for ((tableAlias, column) in spec.columns) {
            if (tableAlias == alias) {
                val exists =
                    findTable(schema, tableSchema, tableName)?.columns?.any { it.name == column }
                        ?: false
                if (!exists) {
                    return Outcome.err("Column '$alias.$column' does not exist")
                }
            }
        }
    }

    for ((tableAlias, _) in spec.columns) {
        if (!tableAliases.contains(tableAlias)) {
            return Outcome.err("Column selection references unknown table alias '$tableAlias'")
        }
    }

    val groupedColumns = mutableSetOf<Pair<String, String>>()
    for ((tableAlias, column) in spec.groups) {
        if (!tableAliases.contains(tableAlias)) {
            return Outcome.err("Group references unknown table alias '$tableAlias'")
        }
        val table =
            findTableByAlias(schema, spec, tableAlias)
                ?: return Outcome.err("Cannot resolve table for group alias '$tableAlias'")
        val groupColumn =
            table.columns.find { it.name == column }
                ?: return Outcome.err("Group column '$tableAlias.$column' does not exist")
        if (dialect != null && !supportsGroupingAndSorting(groupColumn.dataType, dialect)) {
            return Outcome.err(
                "Group column '$tableAlias.$column' has type '${groupColumn.dataType}', " +
                    "which cannot be grouped by ${dialect.displayName()}"
            )
        }
        if (!groupedColumns.add(tableAlias to column)) {
            return Outcome.err("Group column '$tableAlias.$column' is duplicated")
        }
    }

    val sortedColumns = mutableSetOf<Pair<String, String>>()
    for ((tableAlias, column, _) in spec.sorts) {
        if (!tableAliases.contains(tableAlias)) {
            return Outcome.err("Sort references unknown table alias '$tableAlias'")
        }
        val table =
            findTableByAlias(schema, spec, tableAlias)
                ?: return Outcome.err("Cannot resolve table for sort alias '$tableAlias'")
        val sortColumn =
            table.columns.find { it.name == column }
                ?: return Outcome.err("Sort column '$tableAlias.$column' does not exist")
        if (dialect != null && !supportsGroupingAndSorting(sortColumn.dataType, dialect)) {
            return Outcome.err(
                "Sort column '$tableAlias.$column' has type '${sortColumn.dataType}', " +
                    "which cannot be sorted by ${dialect.displayName()}"
            )
        }
        if (!sortedColumns.add(tableAlias to column)) {
            return Outcome.err("Sort column '$tableAlias.$column' is duplicated")
        }
    }

    val distinctSortConflicts = distinctSortProjectionConflicts(spec)
    if (distinctSortConflicts.isNotEmpty()) {
        val labels = distinctSortConflicts.joinToString { "'${it.tableAlias}.${it.column}'" }
        val subject =
            if (distinctSortConflicts.size == 1) "Sort column $labels" else "Sort columns $labels"
        return Outcome.err("$subject must be explicitly selected when Distinct rows is enabled")
    }

    if (spec.groups.isNotEmpty()) {
        if (spec.columns.isEmpty()) {
            return Outcome.err("Grouping requires explicitly selected output columns")
        }
        for ((tableAlias, column) in spec.columns) {
            if ((tableAlias to column) !in groupedColumns) {
                return Outcome.err(
                    "Selected output column '$tableAlias.$column' must appear in GROUP BY"
                )
            }
        }
        for ((tableAlias, column, _) in spec.sorts) {
            if ((tableAlias to column) !in groupedColumns) {
                return Outcome.err("Sort column '$tableAlias.$column' must appear in GROUP BY")
            }
        }
    }

    for (join in spec.joins) {
        if (!tableAliases.contains(join.leftAlias)) {
            return Outcome.err("Join references unknown table alias '${join.leftAlias}'")
        }
        if (!tableAliases.contains(join.rightAlias)) {
            return Outcome.err("Join references unknown table alias '${join.rightAlias}'")
        }

        val leftTable =
            findTableByAlias(schema, spec, join.leftAlias)
                ?: return Outcome.err("Cannot resolve table for alias '${join.leftAlias}'")
        val rightTable =
            findTableByAlias(schema, spec, join.rightAlias)
                ?: return Outcome.err("Cannot resolve table for alias '${join.rightAlias}'")

        val leftCol =
            leftTable.columns.find { it.name == join.leftColumn }
                ?: return Outcome.err(
                    "Join column '${join.leftAlias}.${join.leftColumn}' does not exist"
                )
        val rightCol =
            rightTable.columns.find { it.name == join.rightColumn }
                ?: return Outcome.err(
                    "Join column '${join.rightAlias}.${join.rightColumn}' does not exist"
                )

        val completeForeignKeyJoin = isPartOfCompleteForeignKey(schema, spec, join)
        if (!leftCol.joinEligible && !completeForeignKeyJoin) {
            return Outcome.err(
                "Join column '${join.leftAlias}.${join.leftColumn}' is not the leading key of an equality-capable index"
            )
        }
        if (!rightCol.joinEligible && !completeForeignKeyJoin) {
            return Outcome.err(
                "Join column '${join.rightAlias}.${join.rightColumn}' is not the leading key of an equality-capable index"
            )
        }
        if (leftCol.category != rightCol.category) {
            return Outcome.err(
                "Join columns '${join.leftAlias}.${join.leftColumn}' and '${join.rightAlias}.${join.rightColumn}' have incompatible types"
            )
        }
    }

    when (
        val result =
            validateNode(
                node = FilterNode.Group(spec.filters),
                schema = schema,
                spec = spec,
                tableAliases = tableAliases,
                depth = 0,
                warnings = warnings,
            )
    ) {
        is Outcome.Ok -> Unit
        is Outcome.Err -> return Outcome.err(result.message)
    }

    if (spec.tables.size > 1 && !checkJoinConnectivity(spec)) {
        return Outcome.err("Not all tables are connected by joins — add joins linking every table")
    }

    var limit = spec.limit
    if (limit == 0) {
        limit = DEFAULT_LIMIT
        warnings.add("Limit was 0; defaulted to $DEFAULT_LIMIT")
    }
    if (limit > MAX_LIMIT) {
        warnings.add("Limit $limit exceeds maximum $MAX_LIMIT; capped to $MAX_LIMIT")
        limit = MAX_LIMIT
    }
    if (limit > LARGE_LIMIT_WARNING_THRESHOLD) {
        warnings.add(
            "Large result limits are useful for reporting, but filters, selected columns, and indexed predicates keep queries faster and easier to reuse."
        )
    }

    if (spec.columns.isEmpty()) {
        warnings.add("No columns selected — query will select all columns")
    }

    val normalizedSpec = spec.copy(limit = limit)

    return Outcome.ok(
        normalizedSpec to ValidationOutcome(warnings = warnings.toList(), limit = limit)
    )
}

internal fun findTable(schema: Schema, tableSchema: String, tableName: String) =
    schema.tables.find { it.schema == tableSchema && it.name == tableName }

internal fun findTableByAlias(schema: Schema, spec: QuerySpec, alias: String) =
    spec.tables
        .find { it.alias == alias }
        ?.let { tableRef -> findTable(schema, tableRef.schema, tableRef.name) }

internal fun supportsGroupingAndSorting(dataType: String, dialect: Dialect): Boolean {
    val normalized = dataType.lowercase().substringBefore('(').trim()
    return when (dialect) {
        Dialect.Postgres -> normalized !in setOf("json", "xml")
        Dialect.Mssql ->
            normalized !in setOf("text", "ntext", "image", "xml", "geometry", "geography")
        Dialect.Oracle -> normalized !in setOf("blob", "clob", "nclob", "bfile", "long", "xmltype")
        Dialect.MySql -> true
    }
}

private fun Dialect.displayName(): String =
    when (this) {
        Dialect.Postgres -> "PostgreSQL"
        Dialect.MySql -> "MySQL"
        Dialect.Mssql -> "SQL Server"
        Dialect.Oracle -> "Oracle"
    }

private fun isPartOfCompleteForeignKey(schema: Schema, spec: QuerySpec, join: JoinSpec): Boolean {
    for ((tableSchema, tableName, alias) in spec.tables) {
        val foreignTable = findTable(schema, tableSchema, tableName) ?: continue
        for ((_, columns, referencedSchema, referencedTable, referencedColumns) in
            foreignTable.foreignKeys) {
            if (columns.isEmpty() || columns.size != referencedColumns.size) continue
            val referencedRefs =
                spec.tables.filter { (refSchema, refName, _) ->
                    refSchema == referencedSchema && refName == referencedTable
                }
            for ((_, _, referencedAlias) in referencedRefs) {
                val expected =
                    columns.zip(referencedColumns).map { (foreignColumn, referencedColumn) ->
                        JoinSpec(
                            alias,
                            foreignColumn,
                            referencedAlias,
                            referencedColumn,
                        )
                    }
                if (
                    expected.any { it.matches(join) } &&
                        expected.all { candidate -> spec.joins.any(candidate::matches) }
                ) {
                    return true
                }
            }
        }
    }
    return false
}

private fun JoinSpec.matches(other: JoinSpec): Boolean =
    (leftAlias == other.leftAlias &&
        leftColumn == other.leftColumn &&
        rightAlias == other.rightAlias &&
        rightColumn == other.rightColumn) ||
        (leftAlias == other.rightAlias &&
            leftColumn == other.rightColumn &&
            rightAlias == other.leftAlias &&
            rightColumn == other.leftColumn)
